src/util/assignment.rs

```diff
@@ -0,0 +1,263 @@
+        let mut decompressed_assignment = String::new();
+        decoder.read_to_string(&mut decompressed_assignment)?;
+        let mut result: Assignment = serde_json::from_str(&decompressed_assignment)?;
+        result.id = network_state.assignment.id;
+        Ok(Some(result))
+    }
+    
+    pub fn insert_assignment(&mut self, peer_id: String, status: String, chunks_deltas: Vec<u64>) {
+        self.worker_assignments.insert(peer_id.clone(), WorkerAssignment { 
+            status, 
+            chunks_deltas,
+            encrypted_headers: Default::default()
+        });
+    }
+
+    pub fn dataset_chunks_for_peer_id(&self, peer_id: String) -> Option<Vec<Dataset>> {
+        let local_assignment = match self.worker_assignments.get(&peer_id) {
+            Some(worker_assignment) => worker_assignment,
+            None => {
+                return None
+            }
+        };
+        let mut result: Vec<Dataset> = Default::default();
+        let mut idxs: Vec<u64> = Default::default();
+        let mut cursor = 0;
+        for v in &local_assignment.chunks_deltas {
+            cursor += v;
+            idxs.push(cursor);
+        }
+        cursor = 0;
+        for u in &self.datasets {
+            if idxs.is_empty() {
+                break;
+            }
+            let mut filtered_chunks: Vec<Chunk> = Default::default();
+            for v in &u.chunks {
+                if idxs[0] < cursor {
+                    return None; // Malformed diffs
+                }
+                if idxs[0] == cursor {
+                    filtered_chunks.push(v.clone());
+                    idxs.remove(0);
+                }
+                if idxs.is_empty() {
+                    break;
+                }
+                cursor += 1;
+            }
+            if !filtered_chunks.is_empty() {
+                result.push(Dataset {
+                    id: u.id.clone(),
+                    base_url: u.base_url.clone(),
+                    chunks: filtered_chunks
+                });
+            }
+        }
+        Some(result)
+    }
+
+    pub fn headers_for_peer_id(&self, peer_id: String, secret_key: Vec<u8>) -> Option<BTreeMap<String, String>> {
+        let local_assignment = self.worker_assignments.get(&peer_id)?;
+        let EncryptedHeaders {identity, nonce, ciphertext,} = &local_assignment.encrypted_headers;
+        let Ok(temporary_public_key) = PublicKey::from_slice(identity.as_slice()) else {
+            return None
+        };
+        let big_slice = Sha512::default().chain_update(secret_key).finalize();
+        let Ok(worker_secret_key) = SecretKey::from_slice(&big_slice[00..32]) else {
+            return None
+        };
+        let shared_box = SalsaBox::new(&temporary_public_key, &worker_secret_key);
+        let generic_nonce = GenericArray::clone_from_slice(nonce);
+        let Ok(decrypted_plaintext) = shared_box.decrypt(&generic_nonce, &ciphertext[..]) else {
+            return None
```

Comment:
It should probably be an error. From the function signature my guess is that it returns `None` iff the headers attribute is missing in the network state file for this worker

Reply:
done